lib/srv/desktop/rdp/rdpclient/src/client.rs

```diff
@@ -132,101 +139,125 @@ impl Client {
     /// eventually returns. Failure to do so can result in a leak.
-    fn run_rdp_loop(self) -> oneshot::Receiver<Result<(), ClientError>> {
-        let (result_tx, result_rx) = oneshot::channel::<Result<(), ClientError>>();
+    async fn run_loops(mut self) -> Result<(), ClientError> {
+        let read_stream = self
+            .read_stream
+            .take()
+            .ok_or_else(|| ClientError::InternalError)?;
+
+        let write_stream = self
+            .write_stream
+            .take()
+            .ok_or_else(|| ClientError::InternalError)?;
+
+        let x224_processor = self
+            .x224_processor
+            .take()
+            .ok_or_else(|| ClientError::InternalError)?;
+
+        let write_requester = self
+            .write_requester
+            .take()
+            .ok_or_else(|| ClientError::InternalError)?;
+
+        let write_receiver = self
+            .function_receiver
+            .take()
+            .ok_or_else(|| ClientError::InternalError)?;
+
+        let read_loop_handle = Client::run_read_loop(
+            self.cgo_handle,
+            read_stream,
+            x224_processor,
+            write_requester,
+        );
 
-        global::TOKIO_RT.spawn(async move {
-            let res = self.run_rdp_loop_internal().await;
-            match result_tx.send(res) {
-                Ok(_) => {}
-                Err(res) => {
-                    error!("failed to send result: {:?}", res)
-                }
-            };
-        });
+        let write_loop_handle = Client::run_write_loop(write_stream, write_receiver);
 
-        result_rx
+        // Wait for either loop to finish. When one does, abort the other and return the result.
+        match futures_util::future::try_select(read_loop_handle, write_loop_handle).await {
+            // One of the loops finished successfully. Abort the other and return the result.
+            Ok(either) => match either {
+                futures_util::future::Either::Left((read_loop_res, write_loop_handle)) => {
+                    write_loop_handle.abort();
+                    read_loop_res
+                }
+                futures_util::future::Either::Right((write_loop_res, read_loop_handle)) => {
+                    read_loop_handle.abort();
+                    write_loop_res
+                }
+            },
+            // One of the loops panicked. Abort the other and return the error.
+            Err(either) => match either {
+                futures_util::future::Either::Left((read_loop_panic, write_loop_handle)) => {
+                    write_loop_handle.abort();
+                    Err(read_loop_panic.into())
+                }
+                futures_util::future::Either::Right((write_loop_panic, read_loop_handle)) => {
+                    read_loop_handle.abort();
+                    Err(write_loop_panic.into())
+                }
+            },
+        }
     }
```

Comment:
I think this should do it
```rust
tokio::select! {
    res = &mut read_loop_handle => {
        write_loop_handle.abort();
        res?
    },
    res = &mut write_loop_handle => {
        read_loop_handle.abort();
        res?
    }
}
```

Reply:
Wow, I spent a lot of time trying to get something like that to no avail. Thanks 👍